### DISCUSSION TITLE
[FEATURE] 增加1Panel证书更新后自动重启

### DISCUSSION THREAD

**Discussion by koch3092** [1Panel-dev/1Panel#5113] [Q&A]
### 1Panel 版本

v1.10.7-lts

### 请描述您的需求或者改进建议

面板设置→面板SSL，更新证书后需要重启。

现在日常遇到的情况是：
其实整个过程中，根本就不会去记得更新证书，所以实际上是依赖于网站→证书中的自动更新，但是每三个月更新了证书后，面板的证书又不会自动更新，就还是得手动重启下面板，当需要管理的服务器一多，就感觉每次登录面板，都是https证书过期。

更严重遇到的一个连锁反应是：
如果我设置了HSTS，那从域名进入的方案完全失效。
更更严重的是，如果设置了只允许从域名登录，那理论上是不是只能重装？（没有配置到这么严格过）

### 请描述你建议的实现方案

建议两种方案：
1. 设置一个开关，允许面板SSL证书更新后自动重启服务（当然会有Down掉的风险，可提醒用户注意）。
2. 允许定期重启，比如三个月重启一次，那就跟Let's Encrypt证书的更新时间相同，可减少这种情况。

**Comment by wanghe-fit2cloud** [comment]
感谢反馈。
我们也是考虑到自动重启的风险，才放弃了该策略，改有用户自行控制。
你也可以考虑通过 1Panel 提供的计划任务功能来达到定时重启的效果。

**Comment by koch3092** [comment]
感谢回复。
我不太了解1Panel的整体架构，所以可能会有些理解上的错误。

是否可以局部重启相关的程序，让这个证书完成适配？
之前确实有考虑使用1Panel的计划任务功能来做，不过我不太了解这个计划功能是通过什么级别的限定上下文来运行的，所以我一直没尝试，它是基于Linux自带的Cron来实现的吗？
这个事情我感觉是个对体验上影响还挺大的问题，就有一种，"既然网站都实现了不重启自动适配证书，为啥本身面板的证书无法自适应"的很难受的感觉，而且事实上这个问题对管理的一致性上影响挺大的。

**Reply by wanghe-fit2cloud** [reply]
是通过 1Panel 后端自己的定时来实现的，不依赖 crontab。
网站的证书可以依赖 OpenResty 的重载方法，如果是 1Panel 面板设置中设置的证书，此时需要重启 1Panel 系统来实现的。

**Reply by koch3092** [reply]
那"我重启我自己"这件事情，会有问题吗？如果没啥问题，那就目前只能按照这个方案先弄了

**Reply by wanghe-fit2cloud** [reply]
没有问题的。

**Reply by Sun-drenched** [reply]
注意到7月份更新了“申请证书成功之后支持执行脚本”功能，似乎可以通过这个功能更简便地实现

**Reply by Sun-drenched** [reply]
> 注意到7月份更新了“申请证书成功之后支持执行脚本”功能，似乎可以通过这个功能更简便地实现

不过有一点需要考虑的是自动续签的时间节点，如果存在到期时间相近证书一起发起续签的设定，我在想会不会存在面板证书续签成功后触发自动重启脚本，中断面板其他任务的可能。

**Comment by Ben-Huang** [comment]
相同的问题,之前已经提过 [https://github.com/1Panel-dev/1Panel/discussions/6480](https://github.com/1Panel-dev/1Panel/discussions/6480) 
v1.10.19-lts 复现
```
2024/09/22 00:00:29 开始申请证书
2024/09/22 00:01:47 更新网站证书成功
```
2024/09/22证书已经自动续签
面板 SSL 页面也显示新的证书有限期
![ssl](https://github.com/user-attachments/assets/146bef60-c701-4bc0-a0c7-797c4ab9e9c4)


但是实际使用依然是旧的证书
![llq](https://github.com/user-attachments/assets/17e1a84f-7357-4758-b12f-553d5c785c37)

2024/10/20 当证书到期后就会因为浏览器安全问题无法访问

而且重启面板后问题依旧,依然是旧的证书

**Reply by ssongliu** [reply]
> 相同的问题,之前已经提过 [https://github.com/1Panel-dev/1Panel/discussions/6480](url) v1.10.19-lts 复现
> 
> ```
> 2024/09/22 00:00:29 开始申请证书
> 2024/09/22 00:01:47 更新网站证书成功
> ```
> 
> 2024/09/22证书已经自动续签 面板 SSL 页面也显示新的证书有限期 ![ssl](https://private-user-images.githubusercontent.com/15009374/378196902-146bef60-c701-4bc0-a0c7-797c4ab9e9c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk0OTY3MzgsIm5iZiI6MTcyOTQ5NjQzOCwicGF0aCI6Ii8xNTAwOTM3NC8zNzgxOTY5MDItMTQ2YmVmNjAtYzcwMS00YmMwLWEwYzctNzk3YzRhYjllOWM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIxVDA3NDAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlmZDQxYWZjZDNjMDBmZjVjMmQyNzk5ZDUzNTk5NGUzMzdhZjA4MDYyYTY4NzMxMWEzNjA3MWEzNzViNjBhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3WuEz6rhNjOLqWNYC4z5OlgzCq_BEMs4vKodkJfxtYw)
> 
> 但是实际使用依然是旧的证书 ![llq](https://private-user-images.githubusercontent.com/15009374/378196914-17e1a84f-7357-4758-b12f-553d5c785c37.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk0OTY3MzgsIm5iZiI6MTcyOTQ5NjQzOCwicGF0aCI6Ii8xNTAwOTM3NC8zNzgxOTY5MTQtMTdlMWE4NGYtNzM1Ny00NzU4LWIxMmYtNTUzZDVjNzg1YzM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIxVDA3NDAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4NDgxNzA1YWEwOGZiNmRhZjg5ZDUxZWQxNjliZTljYmUxNmJjNWRkZTY1NTE4MjMzMzBmMDYyN2I5ODRjMzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vu5Ela7Ax5II06Y2gTVm-7kz_PHdELQb9cT05FZCwt8)
> 
> 2024/10/20 当证书到期后就会因为浏览器安全问题无法访问
> 
> 而且重启面板后问题依旧,依然是旧的证书

麻烦加一下我，我看看这个问题
![image](https://github.com/user-attachments/assets/21f0550e-8fe1-4892-b2b4-6f1f76a6b08f)

**Comment by lgc2333** [comment]
> https://github.com/1Panel-dev/1Panel/discussions/6480#discussioncomment-17463492

在对应的证书下勾选 “申请证书之后执行脚本”，填写以下内容
（我是单服务器部署，如果多服务器部署的话看情况调整重启命令）

```bash
systemd-run --on-active=30s /usr/bin/1pctl restart all
```
